test/infrastructure/docker/docker/kind_manager.go

```diff
@@ -83,89 +103,140 @@ func (m *Manager) CreateExternalLoadBalancerNode(name, image, clusterLabel, list
+		// including some ones docker would otherwise do by default.
+		// for now this is what we want. in the future we may revisit this.
+		Privileged:  true,
+		SecurityOpt: []string{"seccomp=unconfined"}, // also ignore seccomp
 		// some k8s things want to read /lib/modules
-		"--volume", "/lib/modules:/lib/modules:ro",
-		"--hostname", name, // make hostname match container name
-		"--network", defaultNetwork,
-		"--name", name, // ... and set the container name
-		// label the node with the cluster ID
-		"--label", clusterLabel,
-		// label the node with the role ID
-		"--label", fmt.Sprintf("%s=%s", nodeRoleLabelKey, role),
+		Binds:       []string{"/lib/modules:/lib/modules:ro"},
+		NetworkMode: defaultNetwork,
+		Tmpfs: map[string]string{
+			"/tmp": "", // various things depend on working /tmp
+			"/run": "", // systemd wants a writable /run
+		},
+		PortBindings: nat.PortMap{},
 	}
+	networkConfig := dockerNetwork.NetworkingConfig{}
 
 	// pass proxy environment variables to be used by node's docker daemon
-	proxyDetails, err := getProxyDetails()
+	envVars := []string{}
+	proxyDetails, err := getProxyDetails(ctx)
 	if err != nil || proxyDetails == nil {
 		return nil, errors.Wrap(err, "proxy setup error")
 	}
 	for key, val := range proxyDetails.Envs {
-		runArgs = append(runArgs, "-e", fmt.Sprintf("%s=%s", key, val))
+		envVars = append(envVars, fmt.Sprintf("%s=%s", key, val))
 	}
+	containerConfig.Env = envVars
 
-	// adds node specific args
-	runArgs = append(runArgs, extraArgs...)
+	configureMounts(opts.Mounts, &hostConfig)
+	configurePortMappings(opts.PortMappings, &hostConfig)
 
-	if usernsRemap() {
+	// Expose the container port if supplied.
+	ports := nat.PortSet{}
+	if opts.Port > 0 {
+		ports[nat.Port(fmt.Sprintf("%d/tcp", opts.Port))] = struct{}{}
+	}
+
+	// We need to make sure any PortMappings are also included in the expose list
+	for port := range hostConfig.PortBindings {
+		ports[port] = struct{}{}
+	}
+	containerConfig.ExposedPorts = ports
+
+	cli, err := types.GetDockerClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "container client error")
+	}
+
+	if usernsRemap(ctx, cli) {
 		// We need this argument in order to make this command work
 		// in systems that have userns-remap enabled on the docker daemon
-		runArgs = append(runArgs, "--userns=host")
+		hostConfig.UsernsMode = "host"
 	}
 
-	if err := run(
-		image,
-		withRunArgs(runArgs...),
-		withMounts(mounts),
-		withPortMappings(portMappings),
-	); err != nil {
-		return nil, err
+	// Make sure we have the image
+	pullResp, err := cli.ImagePull(ctx, opts.Image, dockerTypes.ImagePullOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "image pull error")
 	}
 
-	return types.NewNode(name, image, role), nil
-}
+	// The ImagePull operation may take some time. We need to read the response
+	// until EOF to make sure it is complete before we move on.
+	_, _ = io.Copy(io.Discard, pullResp)
```

Comment:
It might be helpful to write the response into a buffer and also log it for easier debugging in case the containercreate below fails. What do you think?

Reply:
+1